logging/README.md

```diff
@@ -0,0 +1,228 @@
+* `quarkus.google.cloud.logging.structured.mdc.included=[true|false]` - Should the MDC values be included? Defaults to `true`
+* `quarkus.google.cloud.logging.structured.mdc.field-name` - This is the field name to use in ESC for MDC values, defaults to `mdc`
+
+You can use a completely custom structured format by binding a `JsonFormatter` to the CDI context. For example, if you want to filter the 
+included parameters, you could do this: 
+
+```java
+package mypackage;
+
+import java.util.Map;
+import java.util.logging.ErrorManager;
+
+import javax.enterprise.context.ApplicationScoped;
+
+import org.jboss.logmanager.ExtLogRecord;
+
+import io.quarkiverse.googlecloudservices.logging.runtime.JsonFormatter;
+import io.quarkiverse.googlecloudservices.logging.runtime.LoggingConfiguration;
+import io.quarkiverse.googlecloudservices.logging.runtime.TraceInfo;
+import io.quarkiverse.googlecloudservices.logging.runtime.ecs.EscJsonFormat;
+
+@ApplicationScoped
+public class TestFormatter extends EscJsonFormat implements JsonFormatter {
+
+    @Override
+    public void init(LoggingConfiguration config, ErrorManager error) {
+        super.setLoggingConfiguration(config);
+        super.setErrorManager(error);
+    }
+
+    @Override
+    public Map<String, ?> format(ExtLogRecord record, TraceInfo trace) {
+        return super.toEsc(record, trace);
+    }
+
+    @Override
+    protected boolean shouldIncludeParameter(Object p) {
+        return !(p instanceof MyCustomParameterType);
+    }
+}
+```
+
+### Resource type and labels
+You can configure the resource type and the resource labels. This is used by GCP operations as dimensions
+when analysing logs. Please refer to the [GCP documentation](https://cloud.google.com/logging/docs/api/v2/resource-list#resource-types) 
+for the allowable resource types and their associated labels. 
+
+* `quarkus.google.cloud.logging.resource.type` - The type string, defaults to `generic_node`
+* `quarkus.google.cloud.logging.resource.label.<string>` - This is a string map with key/value pairs
+
+### Log record labels
+You can configure default labels for each log record:
+
+* `quarkus.google.cloud.logging.default-label.<string>` - This is a string map with key/value pairs
+
+You can also extract custom labels per record by binding a `LabelExtractor` to the CDI context, e.g.: 
+
+```java
+package mypackage;
+
+import javax.enterprise.context.ApplicationScoped;
+
+import org.jboss.logmanager.ExtLogRecord;
+
+import io.quarkiverse.googlecloudservices.logging.runtime.LabelExtractor;
+
+@ApplicationScoped
+public class TestLabelExtractor implements LabelExtractor {
+
+    public Map<String, String> extract(ExtLogRecord record) {
+        if (record.getParameters() == null) {
+            return null;
+        } else {
+            // do something smart here
+        }
+    }
+}
+```
+
+### Tracing
```

Comment:
Should this be replaced by a support for OpenTracing and OpenTelemetry ?

Reply:
Not really, it would _complement_ OT. The thing we're doing here is to provide the `traceId` as a first class member to Operations Logging (OL). That way, if you use OT with export to Operations Tracing, you can supply the `traceId` here and operations will connect the logs with the trace. In a future version, we could try to do it automagically, but that would require a dependency on OT - so I decided to let the implementer provide the `traceId` manually.

Reply:
(Incidentially, I have the code to integrate OpenTelemetry with Operations Tracing, including picking up the existing `traceId` from the Google load balancers. That's a potential future PR if you're interested.)

Reply:
Usually we check that OT (or open tracing) is present (using build time Capacities) and if so we add the needed code via a build step.

Reply:
Ok, shall we save that for a future improvement?

Reply:
Yes, let's not add things to the already big PR ;)